File: TsinghuaNet.Helper/Models/INetStatus.vb
Imports System.ComponentModel
Imports System.Net.NetworkInformation
Imports MvvmHelpers

Public Enum NetStatus
    Unknown
    Wwan
    Wlan
    Lan
End Enum

Public Interface INetStatus
    Inherits INotifyPropertyChanged
    Property Status As NetStatus
    Property Ssid As String
    Function RefreshAsync() As Task
    Function SuggestAsync() As Task(Of NetState)
End Interface

Public Class NetPingStatus
    Inherits ObservableObject
    Implements INetStatus

    Public Property Status As NetStatus Implements INetStatus.Status
        Get
            Return NetStatus.Unknown
        End Get
        Set(value As NetStatus)

        End Set
    End Property

    Public Property Ssid As String Implements INetStatus.Ssid
        Get
            Return Nothing
        End Get
        Set(value As String)

        End Set
    End Property

    Private Shared Async Function CanConnectTo(uri As String) As Task(Of Boolean)
        Try
            Dim p As New Ping
            Dim reply = Await p.SendPingAsync(uri)
            Return reply.Status = IPStatus.Success
        Catch ex As Exception
            Return False
        End Try
    End Function

    Private Shared Async Function GetSuggestion() As Task(Of NetState)
        If Await CanConnectTo("auth4.tsinghua.edu.cn") Then
            Return NetState.Auth4
        ElseIf Await CanConnectTo("net.tsinghua.edu.cn") Then
            Return NetState.Net
        ElseIf Await CanConnectTo("auth6.tsinghua.edu.cn") Then
            Return NetState.Auth6
        Else
            Return NetState.Unknown
        End If
    End Function

    Public Function RefreshAsync() As Task Implements INetStatus.RefreshAsync
        Return Task.CompletedTask
    End Function

    Public Function SuggestAsync() As Task(Of NetState) Implements INetStatus.SuggestAsync
        Return GetSuggestion()
    End Function
End Class

Public MustInherit Class NetMapStatus
    Inherits ObservableObject
    Implements INetStatus

    Private _Status As NetStatus
    Public Property Status As NetStatus Implements INetStatus.Status
        Get
            Return _Status
        End Get
        Set(value As NetStatus)
            SetProperty(_Status, value)
        End Set
    End Property

    Private _Ssid As String
    Public Property Ssid As String Implements INetStatus.Ssid
        Get
            Return _Ssid
        End Get
        Set(value As String)
            SetProperty(_Ssid, value)
        End Set
    End Property

    Public MustOverride Function RefreshAsync() As Task Implements INetStatus.RefreshAsync

    Private Shared SsidStateMap As New Dictionary(Of String, NetState) From
    {
        {"Tsinghua", NetState.Net},
        {"Tsinghua-5G", NetState.Net},
        {"Tsinghua-IPv4", NetState.Auth4},
        {"Tsinghua-IPv6", NetState.Auth6},
        {"Wifi.郑裕彤讲堂", NetState.Net}
    }

    Public Function SuggestAsync() As Task(Of NetState) Implements INetStatus.SuggestAsync
        Return Task.Run(
            Function()
                Select Case Status
                    Case NetStatus.Lan
                        Return NetState.Auth4
                    Case NetStatus.Wlan
                        If SsidStateMap.ContainsKey(Ssid) Then
                            Return SsidStateMap(Ssid)
                        Else
                            Return NetState.Unknown
                        End If
                    Case Else
                        Return NetState.Unknown
                End Select
            End Function)
    End Function
End Class

File: TsinghuaNet/Model/INetStatus.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Net.NetworkInformation;
using System.Threading.Tasks;
using MvvmHelpers;
using TsinghuaNet.Helper;

namespace TsinghuaNet.Model
{
    public enum NetStatus
    {
        Unknown,
        Wwan,
        Wlan,
        Lan
    }

    public interface INetStatus : INotifyPropertyChanged
    {
        NetStatus Status { get; set; }
        string Ssid { get; set; }
        Task RefreshAsync();
        Task<NetState> SuggestAsync();
    }

    public class NetPingStatus : ObservableObject, INetStatus
    {
        public NetStatus Status
        {
            get => NetStatus.Unknown;
            set { }
        }

        public string Ssid
        {
            get => null;
            set { }
        }

        private static async Task<bool> CanConnectTo(string uri)
        {
            try
            {
                using (Ping p = new Ping())
                {
                    var reply = await p.SendPingAsync(uri);
                    return reply.Status == IPStatus.Success;
                }
            }
            catch (Exception)
            {
                return false;
            }
        }

        private static async Task<NetState> GetSuggestion()
        {
            if (await CanConnectTo("auth4.tsinghua.edu.cn"))
                return NetState.Auth4;
            else if (await CanConnectTo("net.tsinghua.edu.cn"))
                return NetState.Net;
            else if (await CanConnectTo("auth6.tsinghua.edu.cn"))
                return NetState.Auth6;
            else
                return NetState.Unknown;
        }

        public Task RefreshAsync() => Task.CompletedTask;

        public Task<NetState> SuggestAsync() => GetSuggestion();
    }

    public abstract class NetMapStatus : ObservableObject, INetStatus
    {
        private NetStatus status;
        public NetStatus Status
        {
            get => status;
            set => SetProperty(ref status, value);
        }

        private string ssid;
        public string Ssid
        {
            get => ssid;
            set => SetProperty(ref ssid, value);
        }

        public abstract Task RefreshAsync();

        private static readonly Dictionary<string, NetState> SsidStateMap = new Dictionary<string, NetState>()
        {
            ["Tsinghua"] = NetState.Net,
            ["Tsinghua-5G"] = NetState.Net,
            ["Tsinghua-IPv4"] = NetState.Auth4,
            ["Tsinghua-IPv6"] = NetState.Auth6,
            ["Wifi.郑裕彤讲堂"] = NetState.Net
        };

        public Task<NetState> SuggestAsync()
        {
            return Task.Run(() =>
            {
                switch (Status)
                {
                    case NetStatus.Lan:
                        return NetState.Auth4;
                    case NetStatus.Wlan:
                        if (SsidStateMap.ContainsKey(Ssid))
                            return SsidStateMap[Ssid];
                        else
                            return NetState.Unknown;
                    default:
                        return NetState.Unknown;
                }
            });
        }
    }
}
